### PR TITLE
Module: Add Module.Expandable component

### PR DESCRIPTION
**Pull request by galaxy0101**
Follow up from #1240

## Test Plan
manual
unit

### CODE REVIEW

**Comment by galaxy0101** [comment]
Moved unresolved discussions from original thread:
@AlbertCarreras : https://www.loom.com/share/6d309d63f36949f6b6e2f0a5f0365713
https://www.loom.com/share/c98f07484ce84eb0830b3cb40fcf33d9

File: packages/gestalt/src/ExpandableModuleBase.js
Comment (galaxy0101) [inline packages/gestalt/src/ExpandableModuleBase.js:19]: We have "info" and "error" types, we don't have a warning type.

File: packages/gestalt/src/ExpandableModuleBase.js
Comment (galaxy0101) [inline packages/gestalt/src/ExpandableModuleBase.js:79]: fixed the alignment.
<Stack + gap will remove the marginEnd of the last item.

File: packages/gestalt/src/ExpandableModuleBase.js
Comment (galaxy0101) [inline packages/gestalt/src/ExpandableModuleBase.js:37]: move the title icon into 'info', so we will display either info icon or error icon.

File: packages/gestalt/src/ExpandableModuleBase.js
Comment (galaxy0101) [inline packages/gestalt/src/ExpandableModuleBase.js:53]: added accessibilityControls and points to Children component

File: packages/gestalt/src/ExpandableModuleBase.js
Comment (galaxy0101) [inline packages/gestalt/src/ExpandableModuleBase.js:88]: Set aria-roles by adding:
accessibilityControls and accessibilityExpanded in TapArea

File: cypress/integration/accessibility_ExpandableModule_spec.js
Comment (galaxy0101) [inline cypress/integration/accessibility_ExpandableModule_spec.js:1]: added a11y cypress test

Comment (christianvuerings) [reply cypress/integration/accessibility_ExpandableModule_spec.js:1]: Thanks for adding this

File: packages/gestalt/src/ExpandableModule.js
Comment (christianvuerings) [inline packages/gestalt/src/ExpandableModule.js:26]: @galaxy0101 this can now be `<Box rounding={2} borderStyle="shadow">`. We should no longer need any custom CSS for this component.

File: packages/gestalt/src/ExpandableModule.jsdom.test.js
Comment (christianvuerings) [inline packages/gestalt/src/ExpandableModule.jsdom.test.js:14]: @galaxy0101 instead of having to define `accessibilityControls` and `accessibilityLabel` on each item, could we do the following:

```
<ExpandableModule
  id="expandableUniqueID"
  accessibilityExpandLabel="Tap to expand the module"
  accessibilityCollapseLabel="Tap to collapse the module"
  items={[
        title: 'Title1',
        icon: 'lock',
        iconAccessibilityLabel: 'lock icon label',
        summary: ['summary1'],
        children: 'Children1',
  ]}
```

* `id`, `accessibilityExpandLabel` and `accessibilityCollapseLabel` are all required
* We no longer have to set a separate `accessibilityLabel` for each item
* We can dynamically set the `accessibilityControls`, based on the index of the item

Comment (christianvuerings) [reply packages/gestalt/src/ExpandableModule.jsdom.test.js:14]:
```
accessibilityControls=`${id}-${index}`
```

File: packages/gestalt/src/ExpandableModule.js
Comment (galaxy0101) [reply packages/gestalt/src/ExpandableModule.js:26]: Fixed.

File: packages/gestalt/src/ExpandableModule.jsdom.test.js
Comment (AlbertCarreras) [reply packages/gestalt/src/ExpandableModule.jsdom.test.js:14]: I wonder about exposing accessibilityControls  or not. Do screen readers read the id? Does that id provide any information?  I thought about this solution but I wasn't sure if templated ids would not provide enough context. I guess their only function is to link the dependent components, so that might be it.

Comment (galaxy0101) [reply packages/gestalt/src/ExpandableModule.jsdom.test.js:14]: Updated

File: docs/src/ExpandableModule.doc.js
Comment (christianvuerings) [inline docs/src/ExpandableModule.doc.js:58]: @galaxy0101 could we change this to
```
accessibilityExpandLabel="Expand the module"
accessibilityCollapseLabel="Collapse the module"
```

File: docs/src/ExpandableModule.doc.js
Comment (christianvuerings) [inline docs/src/ExpandableModule.doc.js:57]: @galaxy0101 this should be a string `id="ExpandableModuleExample1"`

File: docs/src/ExpandableModule.doc.js
Comment (christianvuerings) [inline docs/src/ExpandableModule.doc.js:82]: @galaxy0101 same here - change it to a string `id="uniqueIDforMultipleItems"`, otherwise we get this error:
image

File: docs/src/ExpandableModule.doc.js
Comment (christianvuerings) [inline docs/src/ExpandableModule.doc.js:83]: @galaxy0101 these comma's are breaking the example:
image

File: packages/gestalt/src/ExpandableModule.flowtest.js
Comment (christianvuerings) [inline packages/gestalt/src/ExpandableModule.flowtest.js:6]: @galaxy0101 this should now break CI, since there is no `id`, `accessibilityExpandLabel` or `accessibilityCollapseLabel` defined on the root component

```
type Props = {|
  id: string,
  accessibilityExpandLabel: string,
  accessibilityCollapseLabel: string,
```

File: docs/src/ExpandableModule.doc.js
Comment (galaxy0101) [reply docs/src/ExpandableModule.doc.js:58]: Fixed.

File: packages/gestalt/src/ExpandableModule.flowtest.js
Comment (galaxy0101) [reply packages/gestalt/src/ExpandableModule.flowtest.js:6]: aha, forgot to run flow test.

File: docs/src/ExpandableModule.doc.js
Comment (galaxy0101) [reply docs/src/ExpandableModule.doc.js:57]: Fixed.

File: docs/src/ExpandableModule.doc.js
Comment (galaxy0101) [reply docs/src/ExpandableModule.doc.js:82]: Fixed.

File: docs/src/ExpandableModule.doc.js
Comment (galaxy0101) [reply docs/src/ExpandableModule.doc.js:83]: Fixed.

**Comment by christianvuerings** [comment]
Looks good from the eng side - @ashleyseto could you take a final look from the design side?

**Comment by ashleyseto** [comment]
Hey this is starting to look really great! I've taken some screenshots of some visual details, and I also have some questions around how much we're implementing in this stage

---

I think we should figure out how to name this component "Module — Expandable" or however we have to write it out. We also have Module that is not expandable that we also need and we'd like these to group together.

---

Screen Shot 2020-10-13 at 11 09 03 AM
I noticed the entire row isn't clickable. It is just this content area so you can end up hover over a good portion of the row and not be able to click it to expand. We want the entire title area to be clickable.

---

Screen Shot 2020-10-13 at 11 09 56 AM
Screen Shot 2020-10-13 at 11 10 05 AM
Some refinement I think around alignment needs to be made. The content in the first line should be center aligned so we don't end up with a perceived large gap at the bottom of each row. We made need to tweak further once we see what it looks like aligned.

---

Screen Shot 2020-10-13 at 11 12 57 AM
Screen Shot 2020-10-13 at 11 19 16 AM
The area between each text block should be 24px when truncation is happening. It looks much smaller right now.

---

Screen Shot 2020-10-13 at 11 13 44 AM
There seems to be a bug with the content on the summary running outside the box. It should truncate before the expand icon.

---

Screen Shot 2020-10-13 at 11 15 43 AM
I was having trouble figuring out if this was the correct line color. We have it specced at lightGray in the design. It looks darker here for some reason.

----

Some other questions

- Are we implementing the expand and collapse animation now?
- Are we implementing the responsive behavior now?

**Comment by galaxy0101** [comment]
Hi, @ashleyseto 
Changes made in this revision
1. Make the entire row clickable
2. center align the title, summary and the arrow.
qq: when there're multiple lines summary, do we center only the first row of summary? @ashleyseto . How do we implement that? @christianvuerings 
3. Fixed the space between title and summary when the content are very long, Truncate properly.
4. Verified the divider line color is lightGray

TBD
1. Name of the component
2. Are we implementing the expand and collapse animation now?
3. Are we implementing the responsive behavior now?

**Comment by galaxy0101** [comment]
Hi, @ashleyseto 
Changes made in this revision
1. Make the entire row clickable
2. center align the title, summary and the arrow.
qq: when there're multiple lines summary, do we center only the first row of summary? @ashleyseto . How do we implement that? @christianvuerings 
3. Fixed the space between title and summary when the content are very long, Truncate properly.
4. Verified the divider line color is lightGray

TBD
1. Name of the component
2. Are we implementing the expand and collapse animation now?
3. Are we implementing the responsive behavior now?

File: packages/gestalt/src/ExpandableModuleBase.js
Comment (galaxy0101) [inline packages/gestalt/src/ExpandableModuleBase.js:82]: Changed back to <Box because it seems <Stack cannot truncate the summary properly.
@christianvuerings

**Comment by ashleyseto** [comment]
> Hi, @ashleyseto
> Changes made in this revision
> 
> 1. Make the entire row clickable
> 2. center align the title, summary and the arrow.
>    qq: when there're multiple lines summary, do we center only the first row of summary? @ashleyseto . How do we implement that? @christianvuerings
> 3. Fixed the space between title and summary when the content are very long, Truncate properly.
> 4. Verified the divider line color is lightGray
> 
> TBD
> 
> 1. Name of the component
> 2. Are we implementing the expand and collapse animation now?
> 3. Are we implementing the responsive behavior now?

Hey,

Sorry for the late reply. Feel free to ping in our slack channel for this component if you don't hear from me.

---
1. Looks good thanks
---
2. I should have been more specific. We want the top row to be baseline aligned. I drew a line here in blue to explain what I meant based on the spec or as close as visually possible.
Screen Shot 2020-10-15 at 1 40 14 PM

---
3. I think there is an extra space happening at the end of the description text line. Can we remove one of them?
Screen Shot 2020-10-15 at 1 36 36 PM
Screen Shot 2020-10-15 at 1 36 42 PM
---
4. Thanks for checking!

---
1. As far as component name, to be consistent, we should go with Module.Expandable like we have for the table.
2. Seems like this will be in a future iteration.
3. Seems like this will be in a future iteration.

**Review by ashleyseto — APPROVED** [review]

**Comment by christianvuerings** [comment]
@galaxy0101 heads up, I fixed the integration test in df33ec6a0a06a280101d672d59f6ed07465d5e17. Should be good to merge after this lands

**Review by christianvuerings — APPROVED** [review]

**Pull request merged**
